optuna/integration/sklearn.py

```diff
@@ -0,0 +1,923 @@
+        timeout=None,  # type: Optional[float]
+        verbose=0  # type: int
+    ):
+        # type: (...) -> None
+
+        _check_sklearn_availability()
+
+        self.cv = cv
+        self.enable_pruning = enable_pruning
+        self.error_score = error_score
+        self.estimator = estimator
+        self.load_if_exists = load_if_exists
+        self.max_iter = max_iter
+        self.n_trials = n_trials
+        self.n_jobs = n_jobs
+        self.param_distributions = param_distributions
+        self.pruner = pruner
+        self.random_state = random_state
+        self.refit = refit
+        self.return_train_score = return_train_score
+        self.sampler = sampler
+        self.scoring = scoring
+        self.storage = storage
+        self.study_name = study_name
+        self.subsample = subsample
+        self.timeout = timeout
+        self.verbose = verbose
+
+    def _check_is_fitted(self):
+        # type: () -> None
+
+        attributes = ['n_splits_', 'scorer_', 'study_']
+
+        if self.refit:
+            attributes += ['best_estimator_', 'refit_time_']
+
+        check_is_fitted(self, attributes)
+
+    def _check_params(self):
+        # type: () -> None
+
+        if not hasattr(self.estimator, 'fit'):
+            raise ValueError(
+                'estimator must be a scikit-learn estimator'
+            )
+
+        if type(self.param_distributions) is not dict:
+            raise ValueError('param_distributions must be a dictionary')
+
+        for name, distribution in self.param_distributions.items():
+            if not isinstance(distribution, distributions.BaseDistribution):
+                raise ValueError(
+                    'value of {} must be a optuna distribution'.format(name)
+                )
+
+        if self.enable_pruning and not hasattr(self.estimator, 'partial_fit'):
+            raise ValueError('estimator must support partial_fit')
+
+        if self.max_iter <= 0:
+            raise ValueError(
+                'max_iter must be > 0, got {}'.format(self.max_iter)
+            )
+
+    def _refit(
+        self,
+        X,  # type: TwoDimArrayType
+        y=None,  # type: Optional[Union[OneDimArrayType, TwoDimArrayType]]
+        **fit_params  # type: Dict[str, Any]
+    ):
+        # type: (...) -> 'OptunaSearchCV'
+
+        n_samples = _num_samples(X)
+
+        self.best_estimator_ = clone(self.estimator)
+
+        try:
+            self.best_estimator_.set_params(**self.study_.best_params)
+        except ValueError as e:
+            logger.exception(e)
+
+        logger.info(
+            'Refitting the estimator using {} samples...'.format(n_samples)
+        )
+
+        start_time = time()
+
+        self.best_estimator_.fit(X, y, **fit_params)
+
+        self.refit_time_ = time() - start_time
+
+        return self
+
+    def _set_verbosity(self):
+        # type: () -> None
+
+        if self.verbose > 1:
+            logging.set_verbosity(DEBUG)
```

Comment:
I'm worried that this method call will change the global log-level of Optuna's logger.
For example, a scenario like the following can happen:
```python
import optuna

# Set loglevel to DEBUG
optuna.logging.set_verbosity(DEBUG)


# Use `OptunaSearchCV`
est = SGDClassifier(max_iter=5, tol=1e-03)
param_dist = {'alpha': distributions.LogUniformDistribution(1e-04, 1e+03)}
optuna_search = optuna.integration.OptunaSearchCV(
        est,
        param_dist,
        verbose=-1  # WARNING
)
optuna_search.fit(X, y)   # Optuna's global log-level has implicitly changed here from DEBUG to WARNING.


# Use Optuna API directly
def objective(trial):
    v = ...
    return v

study = optuna.create_study()

# No DEBUG and INFO log messages won't be outputted in the optimization.
# (This is a surprising behavior for me)
study.optimize(objective, n_trials=100)  
```

Reply:
Yes, the global log-level will be changed certainly. I want to control log messagese with `verbose` like `GridSearchCV`, but I do not know any other way.

Reply:
Hmm.
I also don't know the perfect answer but, IMO, the problem can be mitigated by restoring the old log level before returning from `fit` call.

Reply:
I implemented it as you suggested.